spec/services/shopify_refunder_spec.rb

```diff
@@ -0,0 +1,109 @@
+require 'spec_helper'
+
+require_relative '../../lib/active_merchant/billing/gateways/shopify.rb'
+
+describe ShopifyRefunder do
+  let(:transaction_id) { '0xDEADBEEF' }
+  let(:options) { { reason: 'Actual reason', order_id: transaction_id } }
+  let(:transaction_amount) { 1 }
+  let(:credited_money_in_cents) { 100 }
+  let(:transaction) { double(:transaction, amount: transaction_amount, id: transaction_id) }
+
+  before do
+    allow(ShopifyAPI::Transaction).to receive(:find).and_return(transaction)
+  end
+
+  context '.initialize' do
+    subject { described_class.new(credited_money_in_cents, transaction_id, options) }
+
+    it 'successfully does it\'s thing' do
+      expect(subject).to be_truthy
+    end
+  end
+
+  context '.perform' do
+    let(:pos_refund) { double('refund', errors: []) }
+
+    before do
+      allow(ShopifyAPI::Refund).to receive(:create).and_return(pos_refund)
+    end
+
+    context 'when the shopify transaction is not found' do
+      before do
+        allow(ShopifyAPI::Transaction).to receive(:find).and_return(nil)
+      end
+
+      subject { described_class.new(credited_money_in_cents, transaction_id, options) }
+
+      it 'throws an error' do
+        cause = ->{ subject.perform }
+        expect(&cause).to raise_error(ActiveMerchant::Billing::ShopifyGateway::TransactionNotFoundError)
+      end
+    end
+
+    context 'with a full refund' do
+      subject { described_class.new(credited_money_in_cents, transaction_id, options) }
+
+      it 'performs a refund on shopify' do
+        expect(ShopifyAPI::Refund).to receive(:create)
```

Comment:
I would include all the arguments. One of the cool things about unit tests like this is they reflect the complexity of the object under test. If your tests are complicated and difficult to maintain, it probably means _your object is too_. It's a 🚨 that maybe your object needs some refactoring. In this case, maybe your refunder shouldn't be responsible for building all those arguments? Is there an opportunity to decompose that responsibility into smaller SRP-friendly objects?